File: cookbook/typescript/src/weather_tools.ts
/**
 * Local Tools Example - Everruns SDK
 *
 * Demonstrates client-side tool execution: the agent requests a tool call,
 * the client executes it locally, and sends the result back.
 *
 * Run: npx tsx src/weather_tools.ts
 * Run with verbose: npx tsx src/weather_tools.ts --verbose
 */

import {
  Everruns,
  type ContentPart,
  type ToolCallInfo,
  type ToolDefinition,
  extractToolCalls,
  toolResult,
  toolError,
} from "@everruns/sdk";

const SYSTEM_PROMPT =
  "You are a helpful weather assistant. You have access to a tool called `get_weather` " +
  'that accepts a JSON argument `{"city": "<city name>"}` and returns current weather. ' +
  "When the user asks about weather, call the tool and then summarize the result.";

const WEATHER_DATA: Record<string, [number, string]> = {
  paris: [18, "partly cloudy"],
  tokyo: [22, "sunny"],
  "new york": [15, "rainy"],
};

/** Simulated local weather lookup. */
function getWeather(city: string): Record<string, unknown> {
  const [temp, condition] = WEATHER_DATA[city.toLowerCase()] ?? [20, "clear"];
  return { city, temperature_celsius: temp, condition };
}

/** Dispatch a tool call by name and return a ContentPart with the result. */
function handleToolCall(call: ToolCallInfo): ContentPart {
  if (call.name === "get_weather") {
    const city = (call.arguments.city as string) ?? "unknown";
    return toolResult(call.id, getWeather(city));
  }
  return toolError(call.id, `Unknown tool: ${call.name}`);
}

function weatherTool(): ToolDefinition {
  return {
    type: "client_side",
    name: "get_weather",
    description: "Get current weather for a city.",
    parameters: {
      type: "object",
      properties: {
        city: {
          type: "string",
          description: "City name",
        },
      },
      required: ["city"],
      additionalProperties: false,
    },
  };
}

interface MessageData {
  message?: {
    content?: Array<{ type: string; text?: string }>;
  };
}

function extractText(data: unknown): string | null {
  const d = data as MessageData;
  const content = d?.message?.content;
  if (!Array.isArray(content)) return null;
  const texts = content
    .filter((p) => p.type === "text")
    .map((p) => p.text ?? "");
  return texts.length > 0 ? texts.join("") : null;
}

async function main() {
  const verbose =
    process.argv.includes("--verbose") || process.argv.includes("-v");
  const client = devClient();

  // Create agent with tool-aware system prompt
  const agent = await client.agents.create({
    name: "weather-assistant-ts",
    systemPrompt: SYSTEM_PROMPT,
    tools: [weatherTool()],
  });
  console.log(`Created agent: ${agent.id}`);

  // Create session
  const session = await client.sessions.create({
    agentId: agent.id,
    tools: [weatherTool()],
  });
  console.log(`Created session: ${session.id}\n`);

  const existingEvents = await client.events.list(session.id);
  const baselineEventId = existingEvents[existingEvents.length - 1]?.id;

  const stream = client.events.stream(session.id, {
    maxRetries: 3,
    sinceId: baselineEventId,
  });
  const eventTask = handleEvents(client, session.id, stream, verbose);
  await sleep(250);

  // Send user message
  try {
    await withTimeout(
      client.messages.create(session.id, "What's the weather like in Paris?"),
      30_000,
    );
  } catch (error) {
    if (error instanceof TimeoutError) {
      console.log("Timed out waiting for message submission; ending demo.");
      stream.abort();
      return;
    }
    throw error;
  }

  try {
    await withTimeout(eventTask, 60_000);
  } catch (error) {
    if (error instanceof TimeoutError) {
      console.log("Timed out waiting for turn completion; ending demo.");
      stream.abort();
      return;
    }
    throw error;
  }
}

async function handleEvents(
  client: Everruns,
  sessionId: string,
  stream: AsyncIterable<{ type: string; data: Record<string, unknown> }> & {
    abort: () => void;
  },
  verbose: boolean,
) {
  for await (const event of stream) {
    if (verbose) {
      console.log(
        `\n[EVENT] ${event.type}: ${JSON.stringify(event.data, null, 2)}`,
      );
    }

    switch (event.type) {
      case "tool.call_requested": {
        const toolCalls = extractToolCalls(event.data);
        if (toolCalls.length > 0) {
          console.log(`Agent requested ${toolCalls.length} tool call(s)`);
          const results: ContentPart[] = toolCalls.map((tc) => {
            console.log(
              `  -> Executing ${tc.name}(${JSON.stringify(tc.arguments)})`,
            );
            return handleToolCall(tc);
          });

          // Send tool results back
          await client.messages.createToolResults(sessionId, results);
          console.log("  <- Sent tool results\n");
        }
        break;
      }

      case "output.message.completed": {
        const text = extractText(event.data);
        if (text) {
          console.log(`Assistant: ${text}`);
        }
        break;
      }

      case "turn.completed":
        console.log("\n[Turn completed]");
        stream.abort();
        return;

      case "turn.failed":
        console.log("\n[Turn failed]");
        stream.abort();
        return;
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

class TimeoutError extends Error {}

function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return new Promise((resolve, reject) => {
    const timeout = setTimeout(() => reject(new TimeoutError()), ms);
    promise.then(
      (value) => {
        clearTimeout(timeout);
        resolve(value);
      },
      (error: unknown) => {
        clearTimeout(timeout);
        reject(error);
      },
    );
  });
}

function devClient(): Everruns {
  const apiKey = process.env.EVERRUNS_API_KEY;
  if (!apiKey) {
    throw new Error("EVERRUNS_API_KEY environment variable required");
  }
  const baseUrl = process.env.EVERRUNS_API_URL;
  return new Everruns({ apiKey, baseUrl });
}

main().catch(console.error);

File: cookbook/python/src/weather_tools.py
"""Local Tools Example - Everruns SDK.

Demonstrates client-side tool execution: the agent requests a tool call,
the client executes it locally, and sends the result back.

Run: uv run python src/weather_tools.py
Run with verbose: uv run python src/weather_tools.py --verbose
"""

import asyncio
import json
import os
import sys

from everruns_sdk import ContentPart, Everruns, extract_tool_calls
from everruns_sdk.sse import EventStream, StreamOptions

SYSTEM_PROMPT = (
    "You are a helpful weather assistant. You have access to a tool called `get_weather` "
    'that accepts a JSON argument `{"city": "<city name>"}` and returns current weather. '
    "When the user asks about weather, call the tool and then summarize the result."
)

WEATHER_DATA = {
    "paris": (18, "partly cloudy"),
    "tokyo": (22, "sunny"),
    "new york": (15, "rainy"),
}


def get_weather(city: str) -> dict:
    """Simulated local weather lookup."""
    temp, condition = WEATHER_DATA.get(city.lower(), (20, "clear"))
    return {"city": city, "temperature_celsius": temp, "condition": condition}


def handle_tool_call(call_id: str, name: str, arguments: dict) -> ContentPart:
    """Dispatch a tool call by name and return a ContentPart with the result."""
    if name == "get_weather":
        city = arguments.get("city", "unknown")
        result = get_weather(city)
        return ContentPart.make_tool_result(call_id, result)
    return ContentPart.make_tool_error(call_id, f"Unknown tool: {name}")


def extract_text(data: dict) -> str | None:
    """Extract text content from message event data."""
    message = data.get("message")
    if not message:
        return None
    content = message.get("content")
    if not content or not isinstance(content, list):
        return None
    texts = []
    for part in content:
        if isinstance(part, dict) and part.get("type") == "text":
            text = part.get("text")
            if text:
                texts.append(text)
    return "".join(texts) if texts else None


async def main():
    verbose = "--verbose" in sys.argv or "-v" in sys.argv
    client = dev_client()

    try:
        # Create agent with tool-aware system prompt
        agent = await client.agents.create(
            name="Weather Assistant",
            system_prompt=SYSTEM_PROMPT,
        )
        print(f"Created agent: {agent.id}")

        # Create session
        session = await client.sessions.create(agent_id=agent.id)
        print(f"Created session: {session.id}\n")

        # Send user message
        await client.messages.create(
            session_id=session.id,
            text="What's the weather like in Paris?",
        )

        # Stream events and handle tool calls
        options = StreamOptions(max_retries=3)
        stream = EventStream(client, session.id, options)
        async for event in stream:
            if verbose:
                print(f"\n[EVENT] {event.type}: {json.dumps(event.data, indent=2)}")

            if event.type == "output.message.completed":
                tool_calls = extract_tool_calls(event.data)
                if tool_calls:
                    print(f"Agent requested {len(tool_calls)} tool call(s)")
                    results = []
                    for tc in tool_calls:
                        print(f"  -> Executing {tc.name}({tc.arguments})")
                        results.append(handle_tool_call(tc.id, tc.name, tc.arguments))

                    # Send tool results back
                    await client.messages.create_tool_results(
                        session_id=session.id,
                        results=results,
                    )
                    print("  <- Sent tool results\n")
                else:
                    text = extract_text(event.data)
                    if text:
                        print(f"Assistant: {text}")

            elif event.type == "turn.completed":
                print("\n[Turn completed]")
                break

            elif event.type == "turn.failed":
                print("\n[Turn failed]")
                break

    finally:
        await client.close()


def dev_client() -> Everruns:
    """Create client configured for local development."""
    api_key = os.environ.get("EVERRUNS_API_KEY")
    if not api_key:
        raise ValueError("EVERRUNS_API_KEY environment variable required")
    base_url = os.environ.get("EVERRUNS_API_URL")
    return Everruns(api_key=api_key, base_url=base_url)


if __name__ == "__main__":
    asyncio.run(main())
